test/validation/api/atomic/atomic.c

```diff
@@ -375,46 +371,48 @@ static void test_atomic_xchg_32(void)
 {
 	uint32_t old, new;
 	int i;
-	odp_atomic_u32_t *a32u = &global_mem->a32u;
 	odp_atomic_u32_t *a32u_xchg = &global_mem->a32u_xchg;
+	uint8_t buf[CNT];
+	uint64_t seed = odp_thread_id();
+	uint64_t count_old = 0, count_new = 0;
+
+	odp_random_test_data(buf, CNT, &seed);
 
 	odp_barrier_wait(&global_mem->global_barrier);
 
-	for (i = 0; i < CNT; i++) {
-		new = odp_atomic_fetch_inc_u32(a32u);
+	for (i = 0; i <= CNT; i++) {
+		new = i < CNT ? buf[i] : 0;
 		old = odp_atomic_xchg_u32(a32u_xchg, new);
-
-		if (old & 0x1)
-			odp_atomic_xchg_u32(a32u_xchg, 0);
-		else
-			odp_atomic_xchg_u32(a32u_xchg, 1);
+		count_old += old;
+		count_new += new;
 	}
 
-	odp_atomic_sub_u32(a32u, CNT);
-	odp_atomic_xchg_u32(a32u_xchg, U32_MAGIC);
+	odp_atomic_add_u32(a32u_xchg, count_old);
+	odp_atomic_sub_u32(a32u_xchg, count_new);
 }
 
 static void test_atomic_xchg_64(void)
 {
 	uint64_t old, new;
 	int i;
-	odp_atomic_u64_t *a64u = &global_mem->a64u;
 	odp_atomic_u64_t *a64u_xchg = &global_mem->a64u_xchg;
+	uint8_t buf[CNT];
+	uint64_t seed = odp_thread_id();
+	uint64_t count_old = 0, count_new = 0;
+
+	odp_random_test_data(buf, CNT, &seed);
 
 	odp_barrier_wait(&global_mem->global_barrier);
 
 	for (i = 0; i < CNT; i++) {
```

Comment:
Should the break condition be `i <= CNT` like in `test_atomic_xchg_32()` above?

Reply:
Yes, very good catch. However, the whole setting to zero at the end is unnecessary, so I removed the whole thing and the condition is now i < CNT in both tests.